arrow-array/src/array/union_array.rs

```diff
@@ -241,21 +243,29 @@ impl UnionArray {
     ///
     /// Panics if `index` is greater than the length of the array.
     pub fn type_id(&self, index: usize) -> i8 {
-        assert!(index < self.len());
-        self.data().buffers()[0].as_slice()[self.offset() + index] as i8
+        self.type_ids[index]
+    }
+
+    /// Returns the `type_ids` for this array
```

Comment:
nit to match `offsets` :
```suggestion
    /// Returns the `type_ids` buffer for this array
```